kube play: add exit-code propagation

vrothberg opened the pull request:
PR to implement https://issues.redhat.com/browse/RUN-1776. Please see the individual commits for details.

#### Does this PR introduce a user-facing change?

<!--
If no, just write `None` in the release-note block below. If yes, a release note
is required: Enter your extended release note in the block below. If the PR
requires additional action from users switching to the new release, include the
string "action required".

For more information on release notes, please follow the Kubernetes model:
https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
Add a new `ExitCode` field to Quadlet `.kube` files to allow for reflecting the exit codes of failed containers.
```

vrothberg commented:
Still WIP but in an advanced stage and ready to review and play with it.  Tests are there but I still need to go over the docs.

@ygalblum PTAL

vrothberg commented:
This integration also brings up the question whether Quadlet should force `kube play` to default to a `Restart=Never` for the containers.  Currently it's `always` which is compatible with K8s but I see little value when running in systemd.

Not needed in this PR but something to consider before Quadlet goes GA.

vrothberg commented:
@edsantiago PTAL at the tests

Luap99 commented on libpod/boltdb_state.go:
```diff
@@ -1472,7 +1473,17 @@ func (s *BoltState) PruneContainerExitCodes() error {
+		}
+
 		return timeStampBucket.ForEach(func(rawID, rawTimeStamp []byte) error {
+			if ctrsBucket.Bucket(rawID) != nil {
+				// If the container still exists, don't prune
+				// its exit code since we may still need it.
+				return nil
+			}
```
This seems like a bad idea? This could cause us to grow this table infinite. Consider the common use case where people just do podman start/stop and not recreate the container every time.

vrothberg replied:
I don't think it can grow indefinitely.  There will always only be 1 exit code per container, it's not a growing list or something like that.

The idea here is to just not delete this 1 exit code prematurely (since we need it for the exit-code policies to work).

Luap99 replied:
oh I see, I thought the db was keyed `ts -> code` but it is `id -> code` and `id -> ts` so we can only ever store one. In this case this is fine.

Luap99 commented on libpod/sqlite_state.go:
```diff
@@ -963,7 +964,7 @@ func (s *SQLiteState) PruneContainerExitCodes() (defErr error) {
 	}()
 
-	if _, err := tx.Exec("DELETE FROM ContainerExitCode WHERE (Timestamp <= ?);", fiveMinsAgo); err != nil {
+	if _, err := tx.Exec("DELETE FROM ContainerExitCode WHERE (Timestamp <= ?) AND (ID NOT IN (SELECT ID FROM CotainerConfig))", fiveMinsAgo); err != nil {
```
same here as for the boltdb comment.

Luap99 commented on libpod/service.go:
```diff
@@ -240,8 +307,8 @@ func (p *Pod) maybeRemoveServiceContainer() error {
 		}
-		timeout := uint(0)
 		logrus.Debugf("Removing service container %s", serviceCtr.ID())
+		timeout := uint(10)
```
This is not quite correct I think, I don't think we allow setting the stop timeout for the service container but this should really just pass nil so the stop code will default to `StopTimeout()` which then defaults to 10.

vrothberg replied:
Agreed.  Will update with the next push but will wait for CI feedback.

vrothberg commented on test/system/252-quadlet.bats:
```diff
@@ -637,4 +637,64 @@ EOF
+      # For .kube files we probably need a similar trick as in .container files
+      # and use an ExecStopPost which _is_ being executed in case of failure.
+      service_cleanup $QUADLET_SERVICE_NAME $service_state
+      run_podman '?' kube down $yaml_file
```
If tests are happy, I will also add a commit to fix this fart.

vrothberg replied:
@Luap99 cool with the rest of the changes?

ygalblum commented on libpod/service.go:
```diff
@@ -59,52 +60,84 @@ func (c *Container) IsService() bool {
+				if exitCode != 0 {
+					report.failedContainers++
+				}
+				if !pc.IsInfra() {
```
Shouldn't this condition wrap also getting `exitCode` and  counting `failedContainers`?

edsantiago commented on test/system/260-sdnotify.bats:
```diff
@@ -371,4 +371,100 @@ READY=1" "sdnotify sent MAINPID and READY"
+    fname="$PODMAN_TMPDIR/test.yaml"
+
+    # Create a test matrix with the following arguments:
+    # 1. Exit policy
+    # 2. Command of container 1
+    # 3. Command of container 2
+    # 4. Expected exit code of the service container
```
My old brain doesn't mix vertical and horizontal so well. What would you think of:
```
    #   exit policy | ctr1 command | ctr2 command | service container exit code
```

edsantiago commented on test/system/260-sdnotify.bats:
```diff
@@ -371,4 +371,100 @@ READY=1" "sdnotify sent MAINPID and READY"
+            generate_exit_code_yaml $fname $cmd1 $cmd2 $sdnotify_policy
+            yaml_sha=$(sha256sum $fname)
+            service_container="${yaml_sha:0:12}-service"
+            podman_exit=0
+            if [[ $sdnotify_policy != "ignore" ]];then
+                 podman_exit=$exit_code
+            fi
```
My stupid old brain again. It just doesn't like negative tests. I might find this more readable
```bash
    podman_exit=$exit_code
    if [[ $sdnotify_policy == "ignore" ]]; then
        podman_exit=0
    fi
```

ygalblum commented on test/e2e/quadlet/ports.kube:
```diff
@@ -1,6 +1,9 @@
 [Kube]
 Yaml=/opt/k8s/deployment.yml
 
+## assert-podman-args "--annotation=io.podman.kube.exit.policy=all"
```
Please create separate quadlet files (and test calls) for these checks

edsantiago reviewed (commented):
Very quick feedback because I saw that you're waiting to re-push.

Overall looks great, and the added test time seems very much appropriate for the coverage it gives. Thank you! I will start on review pass 2 now.

ygalblum commented on test/system/260-sdnotify.bats:
```diff
@@ -371,4 +371,100 @@ READY=1" "sdnotify sent MAINPID and READY"
+}
+
+@test "podman kube play - exit-code policy" {
+    fname="$PODMAN_TMPDIR/test.yaml"
```
It's better to use a random file name to avoid possible issues when running locally

ygalblum reviewed (commented):
I'm not sure what I think about not having a simple argument for `kube play`.

vrothberg commented:
> I'm not sure what I think about not having a simple argument for kube play.

Are you referring to the "trick" of using an annotation?  I figured it's cleaner to not expose a flag as it's not something users should use directly but only via Quadlet.

vrothberg commented:
Ready for another/final round of reviews :checkered_flag:

vrothberg commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -481,15 +481,16 @@ There is only one required key, `Yaml`, which defines the path to the Kubernetes
+| ExitPolicy=any                    | --annotation=io.podman.kube.exit.policy=any |
+| LogDriver=journald                | --log-driver journald                       |
+| Network=host                      | --net host                                  |
+| PodmanArgs=--annotation=key=value | --annotation=key=value                      |
```
Great catch, thank you!

ygalblum replied:
@edsantiago QQ, when I added this line, I first copied it from [here](https://github.com/containers/podman/blob/main/docs/source/markdown/podman-systemd.unit.5.md?plain=1#L122) and changed it. What's the difference? Is it the multiple `=` signs?

edsantiago replied:
@ygalblum it's a bug in `go-md2man`, that generates broken man pages. Type `man podman-systemd.unit` on any system right now, search for `HealthStar`, and see the broken table layout. See #18678 for a pretty lame attempt to sweep the issue under the rug.

edsantiago commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -501,6 +502,15 @@ it may be absolute or relative to the location of the unit file.
 
 This key may be used multiple times
 
+### `ExitPolicy=`
+
+Control how the main PID of the systemd service should exit.  The following values are supported:
+- `all`: exit non-zero if all containers have failed (i.e., exited non-zero)
+- `any`: exit non-zero if any container has failed
+- `none`: exit zero and ignore failed containers
```
It may be too late for this, or there may be reasons why it's important to leave it this way, but I need to speak up: this name can be ambiguous and confusing. As a naïve human, I can interpret "ExitPolicy" as "exit success" or "exit failure" or simply "exit", and all/any/none do not give me an intuitive clue. I would strongly prefer a name that makes it 100% clear that this policy refers to the propagation of errors. Unfortunately I can't think of a good example, because naming is always the hardest part of programming. Naming is also the most permanent and most costly in the long term, so if anyone else agrees with my concern, can we brainstorm a little?

TomSweeneyRedHat replied:
Just being cheap and using the [thesaurus](https://www.thesaurus.com/browse/policy), maybe Action, Behavior, Method, Protocol, Strategy?

vrothberg replied:
I think `ExitCodePropagation`  matches closely what @edsantiago is looking for. @ygalblum WDYT?

ygalblum replied:
Yes @vrothberg, I like your suggestion.

TomSweeneyRedHat commented on docs/source/markdown/podman-systemd.unit.5.md:
```diff
@@ -481,15 +481,16 @@ There is only one required key, `Yaml`, which defines the path to the Kubernetes
+| **[Kube] options**                | **podman kube play equivalent**             |
+| --------------------------------- | ------------------------------------------- |
+| ConfigMap=/tmp/config.map         | --config-map /tmp/config.map                |
+| ExitPolicy=any                    | --annotation=io.podman.kube.exit.policy=any |
```
Not for this PR, just something to consider, it would be nice from a user standpoint if we could shorten this and drop `io.podman.kube`, or at the very least drop `io.podman`.  I don't know if that's feasible or not.  But something like:

```suggestion
| ExitPolicy=any                    | --annotation=kube.exit.policy=any |
```

Again, not for this PR, just something to ruminate upon.

ygalblum replied:
Actually, according at the [K8S specification for annotation](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/#syntax-and-character-set), the key name may be divided into a prefix, an FQDN denoting the name of the service that will use this annotation, and name, representing the specific key's name. So, I would change it to `io.podman/kube-exit-policy`

vrothberg replied:
There is large number of other annotations using the format of `io.podman.foo.bar`.  I figured to be consistent.  I'll check what to do with annotation vs. CLI flag etc.

ygalblum commented:
> > I'm not sure what I think about not having a simple argument for kube play.
> 
> Are you referring to the "trick" of using an annotation? I figured it's cleaner to not expose a flag as it's not something users should use directly but only via Quadlet.

Yes, that's what I meant. It's a nice trick for sure. But, now, it means that users can pass it in the YAML file. So, it brings up new issues:
1. You need to test conflicts between the YAML and the argument
2. What about `kube generate` should it include the annotation now?

vrothberg commented on libpod/service.go:
```diff
@@ -59,52 +60,83 @@ func (c *Container) IsService() bool {
+				if err != nil {
+					return nil, err
+				}
+				if pc.IsInfra() {
```
No worries, that was me rushing the fix through :) Thanks for checking!

vrothberg commented:
OK, all comments should be addressed now:
* Moved from an annotation to a proper field in the container config
* Added a new hidden --service-exit-code-propagation flag
* Fixed the formatting issue in the docs

@TomSweeneyRedHat @edsantiago @ygalblum PTAL

edsantiago commented:
@vrothberg in case you haven't seen it, CI is failing, man-page inconsistency

vrothberg commented on docs/source/markdown/podman-container-inspect.1.md.in:
```diff
@@ -20,46 +20,47 @@ The keys of the returned JSON can be used as the values for the --format flag (s
+| .ImageName               | Container image name (string)                      |
+| .IsInfra                 | Is this an infra container? (string: true/false)   |
+| .IsService               | Is this a service container? (string: true/false)  |
+| .KubeExitCodePropagation | Service conatiner's Exit-code propagation (string) |
```
D'oh, typo!

edsantiago replied:
Argh. I need to learn not to rely (only) on `range-diff`

vrothberg replied:
https://twitter.com/vlntnrthbrg/status/1481661376095469570

:-)

edsantiago commented:
LGTM

Luap99 commented:
/lgtm
